cfg/optimize_test.go

```diff
@@ -0,0 +1,531 @@
+			w.Header().Set("Metadata-Flavor", "Google")
+			fmt.Fprint(w, "zones/us-central1-a/machineTypes/n1-standard-1")
+		}
+	})
+	defer closeTestServer(t, server)
+
+	// Override metadataEndpoints for testing.
+	metadataEndpoints = []string{server.URL}
+
+	machineType, err := getMachineType(&mockIsValueSet{})
+	if err != nil {
+		t.Fatalf("getMachineType failed: %v", err)
+	}
+	if machineType != "n1-standard-1" {
+		t.Errorf("getMachineType returned unexpected machine type: %s", machineType)
+	}
+}
+func TestOptimize_DisableAutoConfig(t *testing.T) {
+	resetMetadataEndpoints(t)
+	// Create a test server that returns a matching machine type.
+	server := createTestServer(t, func(w http.ResponseWriter, r *http.Request) {
+		w.Header().Set("Metadata-Flavor", "Google")
+		fmt.Fprint(w, "zones/us-central1-a/machineTypes/a3-highgpu-8g")
+	})
+	defer closeTestServer(t, server)
+
+	// Override metadataEndpoints for testing.
+	metadataEndpoints = []string{server.URL}
+
+	cfg := &Config{}
+	isSet := &mockIsValueSet{setFlags: map[string]bool{"disable-autoconfig": true}, boolFlags: map[string]bool{"disable-autoconfig": true}}
+
+	_, err := Optimize(cfg, isSet)
+	if err != nil {
+		t.Fatalf("Optimize failed: %v", err)
+	}
+
+	if cfg.Write.EnableStreamingWrites {
+		t.Errorf("Expected EnableStreamingWrites to be false")
+	}
+	if cfg.MetadataCache.NegativeTtlSecs != 0 {
+		t.Errorf("Expected NegativeTTLSecs to be 0, got %d", cfg.MetadataCache.NegativeTtlSecs)
+	}
+	if cfg.MetadataCache.TtlSecs != 0 {
+		t.Errorf("Expected TTLSecs to be 0, got %d", cfg.MetadataCache.TtlSecs)
+	}
+	if cfg.MetadataCache.StatCacheMaxSizeMb != 0 {
+		t.Errorf("Expected StatCacheMaxSizeMb to be 0, got %d", cfg.MetadataCache.StatCacheMaxSizeMb)
+	}
+	if cfg.MetadataCache.TypeCacheMaxSizeMb != 0 {
+		t.Errorf("Expected TypeCacheMaxSizeMb to be 0, got %d", cfg.MetadataCache.TypeCacheMaxSizeMb)
+	}
+	if cfg.ImplicitDirs {
+		t.Errorf("Expected ImplicitDirs to be false")
+	}
+	if cfg.FileSystem.RenameDirLimit != 0 {
+		t.Errorf("Expected RenameDirLimit to be 0, got %d", cfg.FileSystem.RenameDirLimit)
+	}
+}
+
+func TestApplyMachineTypeOptimizations_MatchingMachineType(t *testing.T) {
+	resetMetadataEndpoints(t)
+	// Create a test server that returns a matching machine type.
+	server := createTestServer(t, func(w http.ResponseWriter, r *http.Request) {
+		w.Header().Set("Metadata-Flavor", "Google")
+		fmt.Fprint(w, "zones/us-central1-a/machineTypes/a3-highgpu-8g")
+	})
+	defer closeTestServer(t, server)
+
+	// Override metadataEndpoints for testing.
+	metadataEndpoints = []string{server.URL}
+
+	config := DefaultOptimizationConfig
+	cfg := &Config{}
+	isSet := &mockIsValueSet{setFlags: map[string]bool{}}
+
+	optimizedFlags, err := ApplyMachineTypeOptimizations(&config, cfg, isSet)
+	if err != nil {
+		t.Fatalf("ApplyMachineTypeOptimizations failed: %v", err)
+	}
+	if len(optimizedFlags) == 0 {
+		t.Errorf("Expected optimization to be applied")
+	}
+	if cfg.MetadataCache.NegativeTtlSecs != 0 {
```

Comment:
Why haven't we added check for streaming-writes? - I have seen inconsistent check - somewhere on all the  7 override configs and somewhere only for 6.